realm/realm-library/src/androidTest/java/io/realm/ManagedOrderedRealmCollectionTests.java

```diff
@@ -0,0 +1,730 @@
+                    Dog dog = realm.createObject(Dog.class);
+                    dog.setName("Dog " + i);
+                    dog.setAge(i);
+                    dogs.add(dog);
+                }
+                realm.commitTransaction();
+                return dogs;
+
+            case REALMRESULTS:
+                for (int i = 0; i < TEST_SIZE; i++) {
+                    Dog dog = realm.createObject(Dog.class);
+                    dog.setAge(i);
+                    dog.setName("Dog " + i);
+                }
+                realm.commitTransaction();
+                return realm.allObjects(Dog.class);
+
+            default:
+                throw new AssertionError("Unknown collection class: " + collectionClass);
+        }
+
+    }
+
+    @Test
+    public void deleteFirstFromRealm_emptyCollection() {
+        OrderedRealmCollection<NullTypes> collection = createEmptyCollection(realm, collectionClass);
+        realm.beginTransaction();
+        collection.deleteFirstFromRealm();
+        realm.commitTransaction();
+        assertEquals(0, collection.size());
+    }
+
+    @Test
+    public void deleteLastFromRealm() {
+        assertEquals(TEST_SIZE - 1, collection.last().getFieldLong());
+        realm.beginTransaction();
+        collection.deleteLastFromRealm();
+        realm.commitTransaction();
+        assertEquals(TEST_SIZE - 1, collection.size());
+        assertEquals(TEST_SIZE - 2, collection.last().getFieldLong());
+    }
+
+    @Test
+    public void deleteLastFromRealm_emptyCollection() {
+        OrderedRealmCollection<NullTypes> collection = createEmptyCollection(realm, collectionClass);
+        realm.beginTransaction();
+        collection.deleteLastFromRealm();
+        realm.commitTransaction();
+        assertEquals(0, collection.size());
+    }
+
+    // Test all methods that mutate data throw correctly if not inside an transaction.
+    // Due to implementation details both UnsupportedOperation and IllegalState is accepted at this level
+    @Test
+    public void mutableMethodsOutsideTransactions() {
+        for (OrderedCollectionMutatorMethod method : OrderedCollectionMutatorMethod.values()) {
+            try {
+                switch (method) {
+                    case DELETE_INDEX: collection.deleteFromRealm(0); break;
+                    case DELETE_FIRST: collection.deleteFirstFromRealm(); break;
+                    case DELETE_LAST: collection.deleteLastFromRealm(); break;
+                    case ADD_INDEX: collection.add(0, new AllJavaTypes()); break;
+                    case ADD_ALL_INDEX: collection.addAll(0, Arrays.asList(new AllJavaTypes())); break;
+                    case SET: collection.set(0, new AllJavaTypes()); break;
+                    case REMOVE_INDEX: collection.remove(0); break;
+                }
+                fail("Unknown method or it failed to throw: " + method);
+            } catch (IllegalStateException ignored) {
```

Comment:
better separate `IllegalStateException` and `unsupportedOperationException`. And aboves are the same ...